parse/parser_test.go

```diff
@@ -17,500 +17,405 @@
+				},
+				{
+					ReturnValue: &ast.Identifier{Value: "b"},
+				},
+			},
 		},
-		inlineExpressionStmtTestCase: struct {
-			expression string
-		}{"function add( 1, 2 )"},
 	}
 
-	l := parse.NewLexer(input)
-	buf := parse.NewTokenBuffer(l)
-	contract, err := parse.Parse(buf)
+	for _, tt := range tests {
+		input := createTestContractCode(tt.contractTmpl)
+		contract, err := parseTestContract(input)
 
-	if err != nil {
-		t.Errorf("parser error: %q", err)
-		t.FailNow()
-	}
-	for _, fn := range contract.Functions {
-		testFunctionLiteral(t, fn, tcs)
-	}
-}
+		if err != nil {
+			t.Errorf("parser error: %q", err)
+			t.FailNow()
+		}
 
-func testFunctionLiteral(t *testing.T, fn *ast.FunctionLiteral, tt parserTestCase) {
-	t.Helper()
-	switch fn.Name.Value {
-	case "testReturnStatement":
-		testReturnStatementFunc(t, fn, tt)
-	case "testAssignStatement":
-		testAssignStatementFunc(t, fn, tt)
-	case "testIfStatement":
-		testIfElseStatementFunc(t, fn, tt)
-	case "testExpressionStatement":
-		testExpressionStatementFunc(t, fn, tt)
-	case "testInlineReturnStatement":
-		testInlineReturnStatementFunc(t, fn, tt)
-	case "testInlineAssignStatement":
-		testInlineAssignStatementFunc(t, fn, tt)
-	case "testInlineConditionStatement":
-		testInlineConditionStatement(t, fn, tt)
-	case "testInlineExpressionStatement":
-		testInlineExpressionStatement(t, fn, tt)
+		for i, fn := range contract.Functions {
+			runReturnStatementTestCases(t, fn, tt.expectedFnHeaders[i], tt.expected[i])
+		}
 	}
 }
 
-func testReturnStatementFunc(t *testing.T, fn *ast.FunctionLiteral, tt parserTestCase) {
-	t.Log("test return statement")
+func runReturnStatementTestCases(t *testing.T, fn *ast.FunctionLiteral, efhs expectedFnHeader, tt ast.ReturnStatement) {
+	t.Logf("test ReturnStatement - [%s]", fn.Name)
 
-	// test testReturn() function body, parameters, return type
-	if fn.ReturnType != ast.VoidType {
-		t.Errorf("testReturn() has wrong return type expected=%v, got=%v",
-			ast.VoidType.String(), fn.ReturnType.String())
-	}
+	chkFnHeader(t, fn, efhs)
 
-	if len(fn.Parameters) != 0 {
-		t.Errorf("testReturn() has wrong parameters length got=%v",
-			len(fn.Parameters))
-	}
-
-	// test testReturn's return statements
-	for i, stmt := range fn.Body.Statements {
+	for _, stmt := range fn.Body.Statements {
 		returnStmt, ok := stmt.(*ast.ReturnStatement)
 		if !ok {
 			t.Errorf("function body stmt is not *ast.ReturnStatement. got=%T", stmt)
 		}
 
-		tc := tt.returnTestCase[i]
-		testExpression(t, returnStmt.ReturnValue, tc.returnValue)
+		testExpression2(t, returnStmt.ReturnValue, tt.ReturnValue)
 	}
 }
 
-func testAssignStatementFunc(t *testing.T, fn *ast.FunctionLiteral, tt parserTestCase) {
-	t.Log("test assign statement")
+// TODO: add test cases
+func TestAssignStatement(t *testing.T) {
+	tests := []struct {
+		contractTmpl      contractTmplData
+		expectedFnHeaders []expectedFnHeader
+		expected          []ast.AssignStatement
+	}{}
```

Comment:
not add test case yet?

Reply:
sorry for confusing you, this pr still WIP.  I want to show you the structure of test cases for feedback